bsve/server/auth.py

```diff
@@ -0,0 +1,119 @@
+
+    @access.public
+    @describeRoute(
+        Description('Authenticate with a BSVE token')
+        .param('apiroot', 'The BSVE server to authenticate against.',
+               required=False, default='https://dev.bsvecosystem.net/api')
+        .notes('Pass your BSVE username and authentication token using '
+               'HTTP Basic Auth.  Returns a cookie containing a valid '
+               'girder token on success.')
+        .errorResponse('Missing Authorization header', 401)
+        .errorResponse('Invalid BSVE login', 403)
+    )
+    def authenticate(self, params):
+        authHeader = cherrypy.request.headers.get('Authorization')
+        if not authHeader or not authHeader[0:6] == 'Basic ':
+            raise RestException('Use HTTP Basic Authentication', 401)
+
+        try:
+            credentials = base64.b64decode(authHeader[6:]).decode('utf8')
+            if ':' not in credentials:
+                raise TypeError
+        except Exception:
+            raise RestException('Invalid HTTP Authorization header', 401)
+
+        email, token = credentials.split(':', 1)
+        data = self._bsveAuth(
+            params.get('apiroot', 'https://dev.bsvecosystem.net/api'),
+            email,
+            token
+        )
+
+        if data is None:
+            raise RestException('Invalid BSVE login', 403)
+
+        User = ModelImporter.model('user')
+        user = User.findOne({'email': email})
+
+        if not user:
+            login = self._generateLogin(email)
+            firstName = data.get('firstName', 'First')
+            lastName = data.get('lastName', 'Last')
+
+            user = User.createUser(
+                login=login,
+                password=None,
+                firstName=firstName,
+                lastName=lastName,
+                email=email
+            )
+
+        setattr(cherrypy.request, 'girderUser', user)
+        token = self.sendAuthTokenCookie(user)
+        user['authToken'] = {
+            'token': token['_id'],
+            'expires': token['expires']
+        }
+
+        return user
+
+    def _bsveAuth(self, apiroot, email, token):
+        """Validate a BSVE authentication token."""
+        data = {
+            'authTicket': token,
+            'data': {
+                'userName': email,
+                'authTicket': token
+            }
+        }
+        headers = {
+            'Content-Type': 'application/json',
+            'harbinger-auth-ticket': token
+        }
+        resp = requests.post(
+            apiroot + '/auth/validatetoken',
+            json=data,
+            headers=headers
+        )
+        if not resp.ok:
+            return None
+        resp = resp.json()
+        if resp['status'] == 1:
+            return resp['data']
+        return None
+
+    def _generateLogin(self, email):
+        """Generate a login name from an email address."""
+        user = email.lower()
+
+        # poor man's escaping, probably not terribly secure
```

Comment:
Can you adjust the comment to mention what you said here?  This risk seems reasonable and low, better than less specific comment about security.